### DISCUSSION TITLE
custom fonts for text elements (e.g. score, custom)

### DISCUSSION THREAD

**Discussion by ghost** [ppy/osu#30790] [Ideas]
currently, certain elements' fonts are unable to deviate from the defaults provided, excluding custom text. 

all text components should be configurable to any font on the user's computer.

ScoreOverlap (and ComboOverlap although currently that element is immovable/unremovable) should also be made adjustable for those elements.

**Comment by bdach** [comment]
This is not (easily) possible. All fonts used by the game are previously rendered into sprite sheets for efficiency. It's not just a question of "pulling a ttf in".

**Reply by Speykious** [reply]
This is something I don't quite understand tbh. Why not just load the few ttfs that osu! uses and then have a dynamic glyph atlas to render text efficiently? From my experience, loading a ttf doesn't take a noticeable amount of time, unless you're trying to rasterize all the glyphs that are in it at once (which would be way heavier for CJK fonts). That would make it easier for osu! to make fonts customizable.

**Reply by rvnexe** [reply]
@Speykious yea, exactly like how stable skin fonts work, but automatically

**Reply by Speykious** [reply]
> @Speykious yea, exactly like how stable skin fonts work, but automatically
> 

I mean... Legacy skins only have number sprites, that's hardly a "font" tbh.

**Reply by bdach** [reply]
> Why not just load the few ttfs that osu! uses and then have a dynamic glyph atlas to render text efficiently?

We already atlas glyphs dynamically by copying them across from the pre-baked sheets into an actual atlas. And we also already have issues wherein the game will have freezes if you happen to scroll to a comment that has a bunch of CJK characters in because the game has to load the baked sprites in on demand. How bad do you think this is going to be if you have it load and rasterize a TTF on top of that?

Granted I'm not sure how good is the efficiency on the glyph fetching but on a quick look I don't see any easy low-hanging wins there.

**Reply by Speykious** [reply]
I did look at how it was doing that earlier today, and scrolled through my whole song select with original metadata, and it filled up 15 1024x1024 textures in a 2D arena-like fashion. Though it didn't have any freezes. I'm kinda surprised that a simple comment filled with CJK characters can create a freeze like that. Do you have a discussion example I could test on?

For a personal project of mine, I've implemented an efficient dynamic glyph atlas with an LRU cache so that glyphs that are no longer used in the current frame can be overwritten. It looks like this:

https://github.com/user-attachments/assets/f5252d3d-9305-4284-865a-dc48668a8fab

That way it only takes the space necessary to display the text that's currently in frame. But also I'm not getting any freezes there...

**Reply by bdach** [reply]
> Do you have a discussion example I could test on?

I don't have one ready for you, but the current worst case is probably like some kanji sequence where each kanji is on a unique sprite page which then leads to needing to load multiple pre-rendered pages to display the text. I guess rasterising would partially solve that problem but I'm not sure how slow rasterisation is going to be compared to texture retrieval. 

Also SSD vs spinning rust may be a concern here. Most devs will have the former but there are still users out there with the latter.

**Comment by rvnexe** [comment]
Since you can import a sprite in the Skin editor, it's slightly possible if you turn off "Show label" option and replace it with a custom sprite.
As for the numbers, I think they can implement a system like Stable, which you can use custom numbers. (not as easy as "pulling a ttf in" yea)
